frontend/src/components/Terminal.tsx

```diff
@@ -22,9 +22,15 @@ class JsonWebsocketAddon {
       }),
     );
     this._socket.addEventListener("message", (event) => {
-      const { message } = JSON.parse(event.data);
-      if (message.action === "terminal") {
-        terminal.write(message.data);
+      const { action, args } = JSON.parse(event.data);
+      if (action === "run") {
+        terminal.writeln(args.command);
+      }
+      if (action === "output") {
+        args.output.split("\n").forEach((line: string) => {
+          terminal.writeln(line);
+        });
```

Comment:
in the future, it might be nice to navigate to this file in the code editor pane

Reply:
good point!